Support custom initializers in NStepRNN

himkt opened the pull request:
I separate this PR from #8438, which aims to support custom initializer in NStepRNN first.

https://github.com/chainer/chainer/pull/8438

himkt commented:
<details>

```
[2019-11-29 16:32:36] ======================================================= FAILURES =======================================================
[2019-11-29 16:32:36]  test_NStepRNN_param_21_{ out_dtype=(), activation='relu', batches=(4, 3, 2), hidden_size=5, in_dtypes=('float16',), input_size=4, n_layers=4}[native:0] 
[2019-11-29 16:32:36] 
[2019-11-29 16:32:36] device = native:0, args = (), kwargs = {}
[2019-11-29 16:32:36] backend_config = <BackendConfig use_chainerx=True chainerx_device='native:0' use_cuda=False cuda_device=None use_cudnn='never' cudnn_deterministic=False autotune=False cudnn_fast_batch_normalization=False use_ideep='never'>
[2019-11-29 16:32:36] obj = <chainer.testing._bundle.TestNStepRNN_param_21_{ out_dtype=(), activation='relu', batches=(4, 3, 2), hidden_size=5, in_dtypes=('float16',), input_size=4, n_layers=4} object at 0x7feac2ddaf60>
[2019-11-29 16:32:36] 
[2019-11-29 16:32:36]     @pytest.mark.parametrize_device(devices)
[2019-11-29 16:32:36]     def entry_func(device, *args, **kwargs):
[2019-11-29 16:32:36]         backend_config = _make_backend_config(device.name)
[2019-11-29 16:32:36]     
[2019-11-29 16:32:36]         # Forward test
[2019-11-29 16:32:36]         obj = cls()
[2019-11-29 16:32:36]         try:
[2019-11-29 16:32:36]             obj.setup(*args, **kwargs)
[2019-11-29 16:32:36]             obj.run_test_forward(backend_config)
[2019-11-29 16:32:36]         finally:
[2019-11-29 16:32:36]             obj.teardown()
[2019-11-29 16:32:36]     
[2019-11-29 16:32:36]         # If this is a NumpyOpTest instance, skip backward/double-backward
[2019-11-29 16:32:36]         # tests if the forward test succeeds with acceptable errors.
[2019-11-29 16:32:36]         if isinstance(obj, NumpyOpTest):
[2019-11-29 16:32:36]             if obj.is_forward_successful_with_accept_errors:
[2019-11-29 16:32:36]                 return  # success with expected errors
[2019-11-29 16:32:36]     
[2019-11-29 16:32:36]         # Backward test
[2019-11-29 16:32:36]         obj = cls()
[2019-11-29 16:32:36]         try:
[2019-11-29 16:32:36]             obj.setup(*args, **kwargs)
[2019-11-29 16:32:36]             obj.run_test_backward(backend_config)
[2019-11-29 16:32:36]         finally:
[2019-11-29 16:32:36]             obj.teardown()
[2019-11-29 16:32:36]     
[2019-11-29 16:32:36]         # Double-backward test
[2019-11-29 16:32:36]         obj = cls()
[2019-11-29 16:32:36]         try:
[2019-11-29 16:32:36]             obj.setup(*args, **kwargs)
[2019-11-29 16:32:36] >           obj.run_test_double_backward(backend_config)
[2019-11-29 16:32:36] 
[2019-11-29 16:32:36] ../tests/chainerx_tests/op_utils.py:374: 
[2019-11-29 16:32:36] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2019-11-29 16:32:36] ../tests/chainerx_tests/op_utils.py:102: in run_test_double_backward
[2019-11-29 16:32:36]     super(OpTest, self).run_test_double_backward(backend_config)
[2019-11-29 16:32:36] ../../../../virtualenv/python3.5.6/lib/python3.5/site-packages/chainer/testing/function_link.py:318: in run_test_double_backward
[2019-11-29 16:32:36]     do_check()
[2019-11-29 16:32:36] ../../../../virtualenv/python3.5.6/lib/python3.5/site-packages/chainer/testing/function_link.py:313: in do_check
[2019-11-29 16:32:36]     **self.check_double_backward_options)
[2019-11-29 16:32:36] /opt/python/3.5.6/lib/python3.5/contextlib.py:77: in __exit__
[2019-11-29 16:32:36]     self.gen.throw(type, value, traceback)
[2019-11-29 16:32:36] ../../../../virtualenv/python3.5.6/lib/python3.5/site-packages/chainer/testing/function_link.py:36: in raise_if_fail
[2019-11-29 16:32:36]     cls.fail(message, e)
[2019-11-29 16:32:36] ../../../../virtualenv/python3.5.6/lib/python3.5/site-packages/chainer/testing/function_link.py:27: in fail
[2019-11-29 16:32:36]     utils._raise_from(cls, message, exc)
[2019-11-29 16:32:36] ../../../../virtualenv/python3.5.6/lib/python3.5/site-packages/chainer/utils/__init__.py:106: in _raise_from
[2019-11-29 16:32:36]     six.raise_from(new_exc.with_traceback(orig_exc.__traceback__), None)
[2019-11-29 16:32:36] <string>:3: in raise_from
[2019-11-29 16:32:36]     ???
[2019-11-29 16:32:36] ../../../../virtualenv/python3.5.6/lib/python3.5/site-packages/chainer/testing/function_link.py:34: in raise_if_fail
[2019-11-29 16:32:36]     yield
[2019-11-29 16:32:36] ../../../../virtualenv/python3.5.6/lib/python3.5/site-packages/chainer/testing/function_link.py:313: in do_check
[2019-11-29 16:32:36]     **self.check_double_backward_options)
[2019-11-29 16:32:36] ../../../../virtualenv/python3.5.6/lib/python3.5/site-packages/chainer/gradient_check.py:1093: in check_double_backward
[2019-11-29 16:32:36]     utils._raise_from(AssertionError, f.getvalue(), e)
[2019-11-29 16:32:36] ../../../../virtualenv/python3.5.6/lib/python3.5/site-packages/chainer/utils/__init__.py:106: in _raise_from
[2019-11-29 16:32:36]     six.raise_from(new_exc.with_traceback(orig_exc.__traceback__), None)
[2019-11-29 16:32:36] <string>:3: in raise_from
[2019-11-29 16:32:36]     ???
[2019-11-29 16:32:36] ../../../../virtualenv/python3.5.6/lib/python3.5/site-packages/chainer/gradient_check.py:1074: in check_double_backward
[2019-11-29 16:32:36]     detect_nondifferentiable=detect_nondifferentiable)
[2019-11-29 16:32:36] ../../../../virtualenv/python3.5.6/lib/python3.5/site-packages/chainer/gradient_check.py:944: in check_backward
[2019-11-29 16:32:36]     detect_nondifferentiable, is_immutable_params=False
[2019-11-29 16:32:36] ../../../../virtualenv/python3.5.6/lib/python3.5/site-packages/chainer/gradient_check.py:463: in run
[2019-11-29 16:32:36]     self._run()
[2019-11-29 16:32:36] ../../../../virtualenv/python3.5.6/lib/python3.5/site-packages/chainer/gradient_check.py:506: in _run
[2019-11-29 16:32:36]     self._compare_gradients(gx_numeric, gx_backward, directions)
[2019-11-29 16:32:36] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2019-11-29 16:32:36] 
[2019-11-29 16:32:36] self = <chainer.gradient_check._CheckBackward object at 0x7feac3c019e8>
[2019-11-29 16:32:36] gx_numeric = array(0.81967573, shape=(), dtype=float64, device='native:0')
[2019-11-29 16:32:36] gx_backward = array(1.09623528, shape=(), dtype=float64, device='native:0')
[2019-11-29 16:32:36] directions = (array([[[-0.00508317, 0.0166011 , -0.01237365, 0.05843152, -0.02649884],
[2019-11-29 16:32:36]         [0.0303226 , 0.07071867, -0.02758518...0.02915952, -0.04637887, -0.05912007, -0.01674939, -0.03754915]], shape=(5, 5), dtype=float64, device='native:0'), ...)
[2019-11-29 16:32:36] 
[2019-11-29 16:32:36]     def _compare_gradients(self, gx_numeric, gx_backward, directions):
[2019-11-29 16:32:36]         atol = self.atol
[2019-11-29 16:32:36]         rtol = self.rtol
[2019-11-29 16:32:36]         # Compare the gradients
[2019-11-29 16:32:36]         try:
[2019-11-29 16:32:36]             testing.assert_allclose(
[2019-11-29 16:32:36]                 gx_numeric, gx_backward, atol=atol, rtol=rtol)
[2019-11-29 16:32:36]         except AssertionError as e:
[2019-11-29 16:32:36]             eps = self.eps
[2019-11-29 16:32:36]             xs = self.xs
[2019-11-29 16:32:36]             gys = self.gys
[2019-11-29 16:32:36]             f = six.StringIO()
[2019-11-29 16:32:36]             f.write('check_backward failed (eps={} atol={} rtol={})\n'.format(
[2019-11-29 16:32:36]                 eps, atol, rtol))
[2019-11-29 16:32:36]             for i, x in enumerate(xs):
[2019-11-29 16:32:36]                 f.write('inputs[{}]:\n'.format(i))
[2019-11-29 16:32:36]                 f.write('{}\n'.format(x))
[2019-11-29 16:32:36]             for i, gy in enumerate(gys):
[2019-11-29 16:32:36]                 f.write('grad_outputs[{}]:\n'.format(i))
[2019-11-29 16:32:36]                 f.write('{}\n'.format(gy))
[2019-11-29 16:32:36]             for i, d in enumerate(directions):
[2019-11-29 16:32:36]                 f.write('directions[{}]:\n'.format(i))
[2019-11-29 16:32:36]                 f.write('{}\n'.format(d))
[2019-11-29 16:32:36]             f.write('gradients (numeric):  {}\n'.format(gx_numeric))
[2019-11-29 16:32:36]             f.write('gradients (backward): {}\n'.format(gx_backward))
[2019-11-29 16:32:36]             f.write('\n')
[2019-11-29 16:32:36]             f.write('x: numeric gradient, y: backward gradient')
[2019-11-29 16:32:36]             f.write(str(e))
[2019-11-29 16:32:36] >           raise AssertionError(f.getvalue())
[2019-11-29 16:32:36] E           chainer.testing.function_link.FunctionTestError: double backward is not implemented correctly
[2019-11-29 16:32:36] E           
[2019-11-29 16:32:36] E           (caused by)
[2019-11-29 16:32:36] E           AssertionError: check_double_backward failed (eps=0.001 atol=0.1 rtol=0.1)
[2019-11-29 16:32:36] E           input[0]:
[2019-11-29 16:32:36] E           array([[[-0.44433594, -0.7915039 , -0.77783203, -0.8071289 , -0.4453125 ],
[2019-11-29 16:32:36] E                   [-0.16125488, 0.17822266, -0.64599609, 0.3005371 , 0.91699219],
[2019-11-29 16:32:36] E                   [-0.08392334, 0.67724609, 0.26391602, 0.08093262, -0.14086914],
[2019-11-29 16:32:36] E                   [0.02192688, 0.013031  , 0.74560547, -0.89794922, 0.54296875]],
[2019-11-29 16:32:36] E           
[2019-11-29 16:32:36] E                  [[0.210083  , -0.75341797, -0.97949219, -0.0639038 , -0.06082153],
[2019-11-29 16:32:36] E                   [-0.23852539, -0.73095703, -0.06118774, 0.77148438, 0.87109375],
[2019-11-29 16:32:36] E                   [-0.32055664, -0.66308594, 0.47143555, 0.46484375, 0.08422852],
[2019-11-29 16:32:36] E                   [0.08605957, -0.26660156, 0.91796875, -0.86132812, -0.02374268]],
[2019-11-29 16:32:36] E           
[2019-11-29 16:32:36] E                  [[0.59863281, -0.05279541, -0.31982422, 0.5258789 , 0.38720703],
[2019-11-29 16:32:36] E                   [-0.87060547, -0.57617188, 0.37475586, -0.69384766, -0.95605469],
[2019-11-29 16:32:36] E                   [0.46704102, 0.7915039 , 0.265625  , 0.61914062, 0.59667969],
[2019-11-29 16:32:36] E                   [-0.26391602, -0.66308594, -0.2758789 , 0.04785156, -0.6508789 ]],
[2019-11-29 16:32:36] E           
[2019-11-29 16:32:36] E                  [[-0.0770874 , -0.68261719, 0.96728516, -0.46435547, 0.32763672],
[2019-11-29 16:32:36] E                   [-0.89404297, 0.74951172, -0.76855469, -0.67773438, -0.68847656],
[2019-11-29 16:32:36] E                   [0.64453125, -0.24230957, -0.05841064, -0.71826172, -0.90039062],
[2019-11-29 16:32:36] E                   [0.48754883, 0.34204102, -0.67138672, 0.02670288, 0.96044922]]], shape=(4, 4, 5), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           input[1]:
[2019-11-29 16:32:36] E           array([[-0.28979492, 0.40185547, -0.87207031, -0.24658203],
[2019-11-29 16:32:36] E                  [0.47363281, 0.34204102, 0.89453125, 0.06573486],
[2019-11-29 16:32:36] E                  [-0.27294922, -0.76367188, -0.87890625, -0.64208984],
[2019-11-29 16:32:36] E                  [0.22717285, -0.52880859, -0.06097412, -0.36083984]], shape=(4, 4), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           input[2]:
[2019-11-29 16:32:36] E           array([[-0.10778809, 0.2421875 , -0.95117188, 0.59082031],
[2019-11-29 16:32:36] E                  [-0.27954102, 0.62402344, 0.23315430, -0.58447266],
[2019-11-29 16:32:36] E                  [-0.69824219, 0.77050781, 0.29614258, 0.31542969]], shape=(3, 4), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           input[3]:
[2019-11-29 16:32:36] E           array([[-0.05514526, -0.26782227, -0.49365234, 0.01779175],
[2019-11-29 16:32:36] E                  [-0.50341797, -0.67089844, 0.1776123 , 0.41503906]], shape=(2, 4), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           input[4]:
[2019-11-29 16:32:36] E           array([[0.18981934, -0.66259766, 0.19165039, -0.6484375 ],
[2019-11-29 16:32:36] E                  [0.94824219, 0.72949219, 0.3239746 , -0.16796875],
[2019-11-29 16:32:36] E                  [0.79736328, 0.17028809, 0.97802734, 0.46704102],
[2019-11-29 16:32:36] E                  [-0.35253906, -0.46069336, 0.08648682, -0.37841797],
[2019-11-29 16:32:36] E                  [-0.72851562, -0.18249512, -0.61230469, 0.71972656]], shape=(5, 4), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           input[5]:
[2019-11-29 16:32:36] E           array([[-0.7524414 , 0.17333984, 0.921875  , -0.69140625, 0.91601562],
[2019-11-29 16:32:36] E                  [0.63623047, -0.85058594, -0.00173283, -0.43408203, 0.25830078],
[2019-11-29 16:32:36] E                  [0.52685547, -0.20947266, -0.27490234, 0.56396484, 0.26660156],
[2019-11-29 16:32:36] E                  [-0.4958496 , 0.49902344, -0.89697266, -0.06192017, -0.98291016],
[2019-11-29 16:32:36] E                  [-0.22070312, 0.44726562, -0.55175781, 0.6977539 , -0.77734375]], shape=(5, 5), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           input[6]:
[2019-11-29 16:32:36] E           array([0.18737793, -0.09234619, 0.8828125 , -0.58447266, -0.88818359], shape=(5,), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           input[7]:
[2019-11-29 16:32:36] E           array([0.74609375, -0.62597656, 0.4921875 , 0.02182007, 0.84423828], shape=(5,), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           input[8]:
[2019-11-29 16:32:36] E           array([[-0.20654297, 0.71777344, 0.60058594, -0.13293457, 0.92871094],
[2019-11-29 16:32:36] E                  [0.44677734, -0.578125  , 0.16674805, 0.48730469, 0.22338867],
[2019-11-29 16:32:36] E                  [0.41137695, 0.47802734, -0.2578125 , -0.93066406, -0.1286621 ],
[2019-11-29 16:32:36] E                  [0.33935547, -0.3203125 , -0.18139648, 0.46923828, -0.75292969],
[2019-11-29 16:32:36] E                  [0.1194458 , 0.98291016, 0.79443359, -0.99072266, 0.61621094]], shape=(5, 5), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           input[9]:
[2019-11-29 16:32:36] E           array([[-0.22375488, 0.95117188, -0.41455078, 0.83496094, 0.56201172],
[2019-11-29 16:32:36] E                  [0.43579102, 0.03723145, 0.1607666 , -0.73730469, -0.46630859],
[2019-11-29 16:32:36] E                  [-0.46704102, -0.63183594, 0.6821289 , -0.35473633, -0.67138672],
[2019-11-29 16:32:36] E                  [-0.60205078, 0.91748047, -0.58740234, -0.08953857, 0.72363281],
[2019-11-29 16:32:36] E                  [-0.13061523, -0.21765137, -0.63769531, -0.15490723, -0.6274414 ]], shape=(5, 5), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           input[10]:
[2019-11-29 16:32:36] E           array([-0.25048828, 0.68505859, 0.38720703, -0.26635742, -0.69677734], shape=(5,), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           input[11]:
[2019-11-29 16:32:36] E           array([0.9453125 , -0.27294922, 0.81835938, -0.58398438, -0.0270996 ], shape=(5,), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           input[12]:
[2019-11-29 16:32:36] E           array([[-0.89453125, 0.875     , -0.08026123, 0.77832031, 0.63623047],
[2019-11-29 16:32:36] E                  [0.19824219, 0.32690430, 0.80322266, -0.90966797, -0.17175293],
[2019-11-29 16:32:36] E                  [0.81982422, -0.66943359, -0.14648438, 0.92089844, -0.95703125],
[2019-11-29 16:32:36] E                  [0.35546875, 0.36474609, -0.3383789 , 0.36499023, 0.4177246 ],
[2019-11-29 16:32:36] E                  [-0.38671875, -0.75537109, -0.01809692, -0.93847656, 0.01316833]], shape=(5, 5), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           input[13]:
[2019-11-29 16:32:36] E           array([[-0.4086914 , 0.11041260, 0.71142578, 0.95800781, -0.45898438],
[2019-11-29 16:32:36] E                  [-0.8461914 , 0.578125  , -0.06414795, -0.1373291 , 0.38476562],
[2019-11-29 16:32:36] E                  [-0.47875977, -0.00465775, -0.12060547, 0.74951172, -0.15222168],
[2019-11-29 16:32:36] E                  [0.52880859, -0.51953125, -0.63134766, 0.33398438, 0.578125  ],
[2019-11-29 16:32:36] E                  [-0.89990234, 0.12597656, -0.11376953, -0.60009766, -0.12115479]], shape=(5, 5), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           input[14]:
[2019-11-29 16:32:36] ...(659 lines skipped)...
[2019-11-29 16:32:36] E                   [0.93164062, 0.55957031, -0.52148438, 0.734375  , 0.61621094],
[2019-11-29 16:32:36] E                   [-0.87255859, -0.53759766, 0.17932129, -0.72509766, 0.35693359]],
[2019-11-29 16:32:36] E           
[2019-11-29 16:32:36] E                  [[0.984375  , -0.42846680, 0.52197266, -0.90673828, -0.33496094],
[2019-11-29 16:32:36] E                   [0.88916016, 0.27294922, 0.20373535, 0.85644531, -0.63671875],
[2019-11-29 16:32:36] E                   [-0.96435547, -0.6196289 , 0.04373169, -0.00835419, 0.60107422],
[2019-11-29 16:32:36] E                   [0.71875   , -0.57421875, -0.12548828, -0.15673828, -0.890625  ]],
[2019-11-29 16:32:36] E           
[2019-11-29 16:32:36] E                  [[-0.97998047, 0.57958984, -0.44946289, 0.43554688, -0.15734863],
[2019-11-29 16:32:36] E                   [-0.7133789 , -0.61474609, -0.37231445, 0.61035156, -0.97460938],
[2019-11-29 16:32:36] E                   [-0.90185547, 0.131958  , 0.37353516, 0.45361328, -0.04061890],
[2019-11-29 16:32:36] E                   [-0.26464844, 0.68017578, -0.0916748 , -0.35717773, -0.81445312]],
[2019-11-29 16:32:36] E           
[2019-11-29 16:32:36] E                  [[-0.87890625, -0.81787109, 0.36547852, 0.36157227, -0.51367188],
[2019-11-29 16:32:36] E                   [0.28100586, -0.8618164 , 0.74560547, -0.78076172, -0.66210938],
[2019-11-29 16:32:36] E                   [-0.06524658, 0.55175781, 0.70898438, -0.57910156, -0.84667969],
[2019-11-29 16:32:36] E                   [0.57763672, 0.0949707 , 0.57226562, 0.84033203, -0.03805542]]], shape=(4, 4, 5), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           grad_outputs[1]:
[2019-11-29 16:32:36] E           array([[-0.08087158, 0.19799805, 0.19860840, 0.00874329],
[2019-11-29 16:32:36] E                  [-0.38623047, 0.08270264, 0.85009766, 0.94091797],
[2019-11-29 16:32:36] E                  [-0.20837402, 0.59765625, 0.27026367, -0.54003906],
[2019-11-29 16:32:36] E                  [-0.89746094, -0.94287109, -0.75439453, -0.55957031]], shape=(4, 4), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           grad_outputs[2]:
[2019-11-29 16:32:36] E           array([[0.65820312, -0.42895508, 0.56201172, 0.00933075],
[2019-11-29 16:32:36] E                  [-0.72314453, 0.55615234, 0.84277344, 0.88623047],
[2019-11-29 16:32:36] E                  [0.40893555, 0.38793945, 0.09307861, -0.2614746 ]], shape=(3, 4), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           grad_outputs[3]:
[2019-11-29 16:32:36] E           array([[0.96484375, -0.86865234, 0.79541016, -0.47216797],
[2019-11-29 16:32:36] E                  [0.14892578, 0.02572632, 0.10894775, 0.29443359]], shape=(2, 4), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           grad_outputs[4]:
[2019-11-29 16:32:36] E           array([[-0.62890625, -0.45605469, -0.703125  , -0.93945312],
[2019-11-29 16:32:36] E                  [0.87841797, -0.30639648, -0.78076172, -0.24328613],
[2019-11-29 16:32:36] E                  [-0.23181152, 0.33081055, -0.51123047, 0.32299805],
[2019-11-29 16:32:36] E                  [-0.80322266, 0.16174316, -0.78613281, 0.09649658],
[2019-11-29 16:32:36] E                  [0.03948975, -0.4074707 , -0.08856201, -0.92285156]], shape=(5, 4), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           grad_outputs[5]:
[2019-11-29 16:32:36] E           array([[0.19799805, -0.99902344, 0.00362778, 0.0034523 , -0.87841797],
[2019-11-29 16:32:36] E                  [0.89990234, 0.21728516, 0.34399414, -0.07446289, 0.40844727],
[2019-11-29 16:32:36] E                  [-0.63769531, 0.29516602, 0.13623047, 0.90820312, 0.59326172],
[2019-11-29 16:32:36] E                  [0.17065430, -0.08929443, 0.47680664, 0.62451172, 0.85449219],
[2019-11-29 16:32:36] E                  [0.65283203, -0.9399414 , 0.54541016, 0.04354858, 0.77099609]], shape=(5, 5), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           grad_outputs[6]:
[2019-11-29 16:32:36] E           array([-0.08288574, 0.08142090, 0.81542969, 0.11260986, -0.62597656], shape=(5,), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           grad_outputs[7]:
[2019-11-29 16:32:36] E           array([0.36230469, -0.97216797, -0.23986816, 0.6171875 , 0.15222168], shape=(5,), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           grad_outputs[8]:
[2019-11-29 16:32:36] E           array([[-0.69628906, -0.34033203, 0.90332031, -0.93945312, 0.55615234],
[2019-11-29 16:32:36] E                  [-0.63720703, -0.5727539 , 0.30517578, -0.71191406, -0.2380371 ],
[2019-11-29 16:32:36] E                  [-0.18847656, 0.31787109, 0.02365112, -0.03482056, 0.04803467],
[2019-11-29 16:32:36] E                  [0.61279297, 0.60888672, -0.70458984, 0.9868164 , -0.86425781],
[2019-11-29 16:32:36] E                  [0.99169922, -0.75292969, -0.87597656, 0.40795898, 0.921875  ]], shape=(5, 5), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           grad_outputs[9]:
[2019-11-29 16:32:36] E           array([[0.4243164 , 0.32617188, -0.73242188, -0.63916016, -0.68652344],
[2019-11-29 16:32:36] E                  [-0.81884766, 0.94384766, 0.01972961, -0.0982666 , -0.34057617],
[2019-11-29 16:32:36] E                  [-0.0129776 , -0.71289062, -0.69824219, 0.63085938, -0.57763672],
[2019-11-29 16:32:36] E                  [-0.82714844, -0.22546387, 0.78662109, -0.14270020, 0.53271484],
[2019-11-29 16:32:36] E                  [0.99072266, -0.95996094, -0.7836914 , 0.38940430, 0.78125   ]], shape=(5, 5), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           grad_outputs[10]:
[2019-11-29 16:32:36] E           array([-0.01504517, -0.859375  , -0.82324219, -0.78759766, -0.59765625], shape=(5,), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           grad_outputs[11]:
[2019-11-29 16:32:36] E           array([0.75488281, 0.88623047, -0.72851562, 0.71728516, 0.51416016], shape=(5,), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           grad_outputs[12]:
[2019-11-29 16:32:36] E           array([[0.37231445, 0.71289062, -0.53808594, 0.03866577, -0.31323242],
[2019-11-29 16:32:36] E                  [0.31762695, -0.43432617, 0.0049324 , -0.39355469, 0.1050415 ],
[2019-11-29 16:32:36] E                  [-0.67773438, 0.13708496, -0.02857971, -0.74511719, 0.08740234],
[2019-11-29 16:32:36] E                  [-0.59912109, 0.34033203, 0.11621094, -0.53515625, 0.0311737 ],
[2019-11-29 16:32:36] E                  [-0.37011719, 0.69335938, -0.1052246 , -0.79980469, 0.80322266]], shape=(5, 5), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           grad_outputs[13]:
[2019-11-29 16:32:36] E           array([[0.71240234, -0.42724609, -0.48193359, -0.8671875 , -0.36474609],
[2019-11-29 16:32:36] E                  [-0.89648438, 0.88818359, 0.43432617, 0.1072998 , -0.28051758],
[2019-11-29 16:32:36] E                  [-0.68164062, -0.1340332 , -0.4411621 , 0.921875  , -0.80371094],
[2019-11-29 16:32:36] E                  [-0.18603516, -0.98339844, 0.13610840, 0.15319824, -0.72558594],
[2019-11-29 16:32:36] E                  [0.34448242, -0.71435547, 0.01846313, -0.26245117, -0.50195312]], shape=(5, 5), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           grad_outputs[14]:
[2019-11-29 16:32:36] E           array([-0.72753906, -0.76123047, -0.89501953, -0.13024902, 0.5415039 ], shape=(5,), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           grad_outputs[15]:
[2019-11-29 16:32:36] E           array([0.70166016, 0.24255371, -0.24023438, 0.35986328, -0.37255859], shape=(5,), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           grad_outputs[16]:
[2019-11-29 16:32:36] E           array([[0.45336914, 0.82910156, -0.81005859, 0.32983398, -0.28613281],
[2019-11-29 16:32:36] E                  [0.52441406, 0.89013672, -0.55273438, -0.8852539 , 0.28564453],
[2019-11-29 16:32:36] E                  [-0.82226562, 0.3359375 , 0.21447754, -0.9609375 , -0.21240234],
[2019-11-29 16:32:36] E                  [-0.08099365, -0.95751953, -0.25268555, 0.13110352, -0.98242188],
[2019-11-29 16:32:36] E                  [-0.6743164 , 0.79638672, -0.2890625 , -0.38037109, 0.16430664]], shape=(5, 5), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           grad_outputs[17]:
[2019-11-29 16:32:36] E           array([[0.04434204, 0.76513672, -0.45532227, -0.40039062, -0.38330078],
[2019-11-29 16:32:36] E                  [-0.79785156, -0.4099121 , 0.45776367, -0.3161621 , -0.42797852],
[2019-11-29 16:32:36] E                  [0.91259766, 0.99902344, -0.32519531, -0.04974365, 0.65917969],
[2019-11-29 16:32:36] E                  [-0.80126953, 0.01623535, 0.66748047, 0.54003906, 0.63183594],
[2019-11-29 16:32:36] E                  [-0.89697266, 0.36254883, 0.92675781, 0.40185547, -0.18029785]], shape=(5, 5), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           grad_outputs[18]:
[2019-11-29 16:32:36] E           array([-0.5859375 , 0.91210938, -0.18798828, 0.43383789, 0.34472656], shape=(5,), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           grad_outputs[19]:
[2019-11-29 16:32:36] E           array([0.68408203, 0.88134766, -0.06039429, -0.78466797, 0.13342285], shape=(5,), dtype=float16, device='native:0')
[2019-11-29 16:32:36] E           directions[0]:
[2019-11-29 16:32:36] E           array([[[-0.00508317, 0.0166011 , -0.01237365, 0.05843152, -0.02649884],
[2019-11-29 16:32:36] E                   [0.0303226 , 0.07071867, -0.02758518, 0.05720378, 0.02944993],
[2019-11-29 16:32:36] E                   [-0.00723459, 0.02109910, -0.00837457, 0.01056028, -0.02593968],
[2019-11-29 16:32:36] E                   [0.05790964, -0.06750449, 0.02173489, -0.05823048, -0.01206672]],
[2019-11-29 16:32:36] E           
[2019-11-29 16:32:36] E                  [[0.0436915 , -0.01582653, 0.01250607, -0.00511661, 0.04797844],
[2019-11-29 16:32:36] E                   [-0.00508546, -0.07677639, -0.03901826, 0.06167248, 0.04210095],
[2019-11-29 16:32:36] E                   [0.02891604, 0.06636082, 0.01806912, 0.04204245, 0.03081963],
[2019-11-29 16:32:36] E                   [-0.02004398, 0.00839584, -0.05972193, 0.06923802, -0.06548286]],
[2019-11-29 16:32:36] E           
[2019-11-29 16:32:36] E                  [[-0.07830632, 0.10529265, -0.01682248, 0.05356744, 0.00462800],
[2019-11-29 16:32:36] E                   [0.10799915, 0.11144009, 0.00467285, 0.01117657, 0.00727667],
[2019-11-29 16:32:36] E                   [0.04415434, 0.01076343, -0.04712688, 0.06080117, -0.02515315],
[2019-11-29 16:32:36] E                   [0.03051392, 0.01084759, 0.1042889 , -0.02068427, -0.08261962]],
[2019-11-29 16:32:36] E           
[2019-11-29 16:32:36] E                  [[-0.04474787, 0.04619136, -0.02974401, -0.04686459, -0.05538337],
[2019-11-29 16:32:36] E                   [0.1079612 , -0.16880846, 0.03493570, -0.02403867, 0.06014997],
[2019-11-29 16:32:36] E                   [0.02387062, 0.01799698, 0.01660836, -0.0481759 , 0.01394483],
[2019-11-29 16:32:36] E                   [0.0045973 , 0.01149023, -0.02128502, -0.03173606, -0.01199959]]], shape=(4, 4, 5), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[1]:
[2019-11-29 16:32:36] E           array([[-0.00463735, 0.01703527, 0.12422940, -0.08352526],
[2019-11-29 16:32:36] E                  [-0.00463036, 0.03165550, 0.03874876, 0.04154721],
[2019-11-29 16:32:36] E                  [0.0121245 , -0.01729964, 0.02033877, 0.01077254],
[2019-11-29 16:32:36] E                  [0.08352639, 0.0058078 , -0.00990822, 0.04723035]], shape=(4, 4), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[2]:
[2019-11-29 16:32:36] E           array([[-0.04789695, -0.00486284, 0.0186626 , -0.05641161],
[2019-11-29 16:32:36] E                  [-0.04086062, -0.02624069, 0.02942448, 0.00458765],
[2019-11-29 16:32:36] E                  [0.03494446, -0.0390188 , 0.03508278, 0.03451449]], shape=(3, 4), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[3]:
[2019-11-29 16:32:36] E           array([[-0.0110307 , 0.02446993, -0.04295531, 0.02373479],
[2019-11-29 16:32:36] E                  [-0.00626834, 0.08405853, 0.09019952, -0.03598728]], shape=(2, 4), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[4]:
[2019-11-29 16:32:36] E           array([[0.06825033, -0.00695168, 0.05213979, 0.0105371 ],
[2019-11-29 16:32:36] E                  [0.0559575 , -0.04999825, 0.04896102, -0.05768313],
[2019-11-29 16:32:36] E                  [-0.02443739, -0.00918767, 0.07997747, -0.02198163],
[2019-11-29 16:32:36] E                  [0.03139115, -0.06381051, 0.01504431, -0.03974665],
[2019-11-29 16:32:36] E                  [-0.08606211, 0.0068682 , 0.05806476, -0.05974569]], shape=(5, 4), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[5]:
[2019-11-29 16:32:36] E           array([[0.0047062 , 0.02986267, 0.03006822, 0.07379315, 0.07352056],
[2019-11-29 16:32:36] E                  [-0.04510317, 0.02697503, -0.02058163, -0.03050406, -0.01712319],
[2019-11-29 16:32:36] E                  [-0.07449502, -0.02290038, 0.03826031, -0.06061518, 0.03976644],
[2019-11-29 16:32:36] E                  [-0.00666849, -0.03084823, 0.05241873, -0.04588199, 0.03594665],
[2019-11-29 16:32:36] E                  [0.02915952, -0.04637887, -0.05912007, -0.01674939, -0.03754915]], shape=(5, 5), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[6]:
[2019-11-29 16:32:36] E           array([-0.01116473, 0.03268710, -0.0154418 , 0.00886985, 0.02181490], shape=(5,), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[7]:
[2019-11-29 16:32:36] E           array([0.00702439, -0.06325917, 0.02485925, -0.0836423 , -0.09608483], shape=(5,), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[8]:
[2019-11-29 16:32:36] E           array([[0.09785104, 0.01881736, 0.01343974, -0.08217206, -0.01417427],
[2019-11-29 16:32:36] E                  [-0.01034955, 0.07797403, -0.02921794, -0.01016152, -0.0050994 ],
[2019-11-29 16:32:36] E                  [0.06008599, -0.03041875, 0.06013548, 0.0418924 , -0.04116919],
[2019-11-29 16:32:36] E                  [0.02026037, 0.02545448, -0.0138671 , -0.02853977, -0.02073203],
[2019-11-29 16:32:36] E                  [-0.05081416, -0.05752755, -0.03135246, 0.02839948, -0.03886031]], shape=(5, 5), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[9]:
[2019-11-29 16:32:36] E           array([[0.0248059 , -0.02596485, -0.02543246, 0.00884921, -0.09933855],
[2019-11-29 16:32:36] E                  [-0.04773314, -0.05794893, -0.00880415, 0.08614709, 0.02741221],
[2019-11-29 16:32:36] E                  [0.02455633, 0.03134388, 0.05070149, 0.07417138, -0.02082517],
[2019-11-29 16:32:36] E                  [-0.03092865, 0.02878142, -0.02419276, 0.01365921, -0.01106097],
[2019-11-29 16:32:36] E                  [-0.0059689 , -0.0053054 , 0.00477580, 0.07612058, -0.0812237 ]], shape=(5, 5), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[10]:
[2019-11-29 16:32:36] E           array([0.05166347, 0.06012649, 0.01118729, -0.01976788, -0.03013477], shape=(5,), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[11]:
[2019-11-29 16:32:36] E           array([-0.04025535, 0.04260022, -0.07657733, 0.04610895, -0.00458496], shape=(5,), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[12]:
[2019-11-29 16:32:36] E           array([[-0.03575327, -0.02859230, -0.08036630, 0.10025956, 0.0524964 ],
[2019-11-29 16:32:36] E                  [0.01506029, -0.02576909, 0.04579416, -0.04622329, -0.03931759],
[2019-11-29 16:32:36] E                  [-0.05397015, -0.02171050, -0.03808014, 0.01896793, 0.06685206],
[2019-11-29 16:32:36] E                  [-0.12087895, -0.01366538, 0.0451878 , 0.0051973 , 0.03991259],
[2019-11-29 16:32:36] E                  [0.0585276 , -0.05818817, 0.014243  , -0.03198954, 0.00967678]], shape=(5, 5), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[13]:
[2019-11-29 16:32:36] E           array([[-0.05184812, 0.0326533 , -0.08336604, 0.01250846, -0.01489008],
[2019-11-29 16:32:36] E                  [0.14654419, -0.09218787, -0.06263536, -0.01812365, -0.01120292],
[2019-11-29 16:32:36] E                  [-0.01985270, 0.01784853, -0.00670923, 0.03830011, -0.03734944],
[2019-11-29 16:32:36] E                  [-0.00641263, -0.0444303 , 0.02994677, 0.0131558 , -0.02291957],
[2019-11-29 16:32:36] E                  [0.06636084, -0.0774055 , -0.02077249, 0.00461569, -0.03584342]], shape=(5, 5), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[14]:
[2019-11-29 16:32:36] E           array([0.04014553, 0.04298036, 0.04301144, 0.00709302, -0.02625173], shape=(5,), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[15]:
[2019-11-29 16:32:36] E           array([0.01531787, -0.00460314, 0.00644599, -0.02639178, -0.00920497], shape=(5,), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[16]:
[2019-11-29 16:32:36] E           array([[-0.01621713, -0.03818810, 0.02311802, 0.02819503, -0.00460774],
[2019-11-29 16:32:36] E                  [0.07832386, 0.06568441, -0.02052548, -0.03225366, -0.00917520],
[2019-11-29 16:32:36] E                  [0.00799473, 0.0671427 , 0.01019308, -0.01576355, 0.0304821 ],
[2019-11-29 16:32:36] E                  [-0.03956738, 0.01583710, -0.05528300, 0.01980395, 0.02103506],
[2019-11-29 16:32:36] E                  [-0.00529907, -0.05197798, 0.00471144, 0.00855928, 0.02978196]], shape=(5, 5), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[17]:
[2019-11-29 16:32:36] E           array([[0.06328829, -0.00926664, 0.04778427, 0.03891742, -0.04449369],
[2019-11-29 16:32:36] E                  [-0.0244965 , 0.03410254, -0.06597792, -0.00584316, 0.01134724],
[2019-11-29 16:32:36] E                  [-0.05841365, 0.04391360, -0.01737157, 0.00575232, 0.02280695],
[2019-11-29 16:32:36] E                  [-0.00739619, -0.03257205, -0.04619883, 0.01466865, 0.05688771],
[2019-11-29 16:32:36] E                  [-0.04324777, -0.02597931, 0.00465366, -0.01681591, 0.01907938]], shape=(5, 5), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[18]:
[2019-11-29 16:32:36] E           array([-0.02023620, 0.0182882 , 0.09853175, -0.03808010, -0.04170825], shape=(5,), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[19]:
[2019-11-29 16:32:36] E           array([0.01990376, -0.02276319, -0.04631708, -0.01148137, -0.00584356], shape=(5,), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[20]:
[2019-11-29 16:32:36] E           array([[[0.05649382, -0.01692454, -0.03836363, -0.05946288, -0.00856958],
[2019-11-29 16:32:36] E                   [0.07068324, 0.00715998, 0.02694668, 0.01857613, 0.02610040],
[2019-11-29 16:32:36] E                   [0.01978947, 0.04167040, -0.04219352, -0.04833599, 0.06072312],
[2019-11-29 16:32:36] E                   [-0.06075679, 0.07177662, -0.08703442, 0.00786955, -0.01024481]],
[2019-11-29 16:32:36] E           
[2019-11-29 16:32:36] E                  [[0.00831589, -0.04192643, 0.05948879, -0.05117262, -0.02715308],
[2019-11-29 16:32:36] E                   [0.01549316, -0.03077907, -0.00469937, -0.02431078, 0.12460192],
[2019-11-29 16:32:36] E                   [0.03878695, 0.02831637, 0.0171185 , -0.00581686, 0.0113719 ],
[2019-11-29 16:32:36] E                   [0.03579426, 0.06396680, 0.02112186, 0.01978215, 0.03448803]],
[2019-11-29 16:32:36] E           
[2019-11-29 16:32:36] E                  [[-0.06955017, 0.03514773, -0.02686388, -0.0787683 , -0.01314888],
[2019-11-29 16:32:36] E                   [-0.00460672, -0.06393169, 0.0294736 , -0.03232090, -0.03938298],
[2019-11-29 16:32:36] E                   [-0.04036468, 0.07691546, 0.01476148, -0.02943679, -0.0102215 ],
[2019-11-29 16:32:36] E                   [0.01605604, 0.00458543, 0.03854729, -0.01786832, -0.00629158]],
[2019-11-29 16:32:36] E           
[2019-11-29 16:32:36] E                  [[-0.08212969, 0.06005919, 0.01170955, -0.00469259, 0.05341929],
[2019-11-29 16:32:36] E                   [-0.00619534, -0.02118437, -0.05749371, -0.04044308, 0.0829223 ],
[2019-11-29 16:32:36] E                   [0.03581049, 0.11285827, -0.05042944, -0.00860718, 0.00742566],
[2019-11-29 16:32:36] E                   [0.0875378 , 0.02010416, -0.00955378, -0.00459246, -0.01887142]]], shape=(4, 4, 5), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[21]:
[2019-11-29 16:32:36] E           array([[0.0250082 , 0.07114296, 0.04357292, 0.03625515, -0.07335839],
[2019-11-29 16:32:36] E                  [0.04874699, -0.09550199, 0.02951905, -0.08770024, 0.11681963],
[2019-11-29 16:32:36] E                  [0.01602124, -0.00471276, 0.10031716, -0.05476875, -0.07172176],
[2019-11-29 16:32:36] E                  [0.00592179, 0.06362205, -0.03012347, 0.01233518, 0.04228387]], shape=(4, 5), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[22]:
[2019-11-29 16:32:36] E           array([[-0.00542188, -0.05140065, -0.0090268 , 0.01168762, 0.02353986],
[2019-11-29 16:32:36] E                  [0.09140717, -0.05360895, 0.04202982, 0.02231397, -0.03574538],
[2019-11-29 16:32:36] E                  [-0.03721863, -0.00746792, -0.00582159, -0.00519411, 0.03851758]], shape=(3, 5), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           directions[23]:
[2019-11-29 16:32:36] E           array([[-0.01123797, -0.03162564, -0.05549029, 0.01442429, -0.02301022],
[2019-11-29 16:32:36] E                  [-0.04699852, -0.03217562, -0.08255585, -0.01356408, -0.08804534]], shape=(2, 5), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           gradients (numeric):  array(0.81967573, shape=(), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           gradients (backward): array(1.09623528, shape=(), dtype=float64, device='native:0')
[2019-11-29 16:32:36] E           
[2019-11-29 16:32:36] E           x: numeric gradient, y: backward gradient
[2019-11-29 16:32:36] E           Not equal to tolerance rtol=0.1, atol=0.1
[2019-11-29 16:32:36] E           
[2019-11-29 16:32:36] E           (mismatch 100.0%)
[2019-11-29 16:32:36] E            x: array(0.819676)
[2019-11-29 16:32:36] E            y: array(1.096235)
[2019-11-29 16:32:36] E           
[2019-11-29 16:32:36] E           assert_allclose failed: 
[2019-11-29 16:32:36] E             shape: () ()
[2019-11-29 16:32:36] E             dtype: float64 float64
[2019-11-29 16:32:36] E             i: (0,)
[2019-11-29 16:32:36] E             x[i]: 0.8196757264305472
[2019-11-29 16:32:36] E             y[i]: 1.0962352756625444
[2019-11-29 16:32:36] E             relative error[i]: 0.25228119854549447
[2019-11-29 16:32:36] E             absolute error[i]: 0.2765595492319972
[2019-11-29 16:32:36] E             relative tolerance * |y[i]|: 0.10962352756625444
[2019-11-29 16:32:36] E             absolute tolerance: 0.1
[2019-11-29 16:32:36] E             total tolerance: 0.20962352756625446
[2019-11-29 16:32:36] E           x: 0.81967573
[2019-11-29 16:32:36] E           y: 1.09623528
[2019-11-29 16:32:36] 
[2019-11-29 16:32:36] ../../../../virtualenv/python3.5.6/lib/python3.5/site-packages/chainer/gradient_check.py:536: FunctionTestError
[2019-11-29 16:32:36] =============================================== short test summary info ================================================
[2019-11-29 16:32:36] FAIL ../tests/chainerx_tests/unit_tests/routines_tests/test_rnn.py::test_NStepRNN_param_21_{ out_dtype=(), activation='relu', batches=(4, 3, 2), hidden_size=5, in_dtypes=('float16',), input_size=4, n_layers=4}[native:0]
[2019-11-29 16:32:36] ================================== 1 failed, 89 passed, 90 skipped in 189.05 seconds ===================================
```

</details>

himkt commented:
> avoid fixing the style in this PR
I'm sorry, it is not intended. I'll revert style changes and make a commit.

himkt commented:
I remove code-fixings.

> Could you keep the default initializer for b?

I'm sorry but could you please elaborate on this?
In my understanding, a bias should be a zero vector by default
because `initial_bias` is zero (default) and `bias_initializer` is `chainer.initializers.constant.Constant` (`initializers._get_initializer(0)`).

```python
initialW, initial_bias = argument.parse_kwargs(
    kwargs,
    ('initialW', None),
    ('initial_bias', None)
# ....
if initial_bias is None:
    initial_bias = 0
bias_initializer = initializers._get_initializer(initial_bias)
```

https://github.com/chainer/chainer/pull/8489/files#diff-e40e91622254e52ad942c6a6d3778b14R71-R73

toslunar commented:
> > Could you keep the default initializer for b?
> 
> I'm sorry but could you please elaborate on this?
> In my understanding, a bias should be a zero vector by default (same as current impl)
> because `initial_bias` is zero (default) and `bias_initializer` is `chainer.initializers.constant.Constant` (`initializers._get_initializer(0)`).
> 
> ```python
> initialW, initial_bias = argument.parse_kwargs(
>     kwargs,
>     ('initialW', None),
>     ('initial_bias', None)
> # ....
> if initial_bias is None:
>     initial_bias = 0
> bias_initializer = initializers._get_initializer(initial_bias)
> ```
> 
> https://github.com/chainer/chainer/pull/8489/files#diff-e40e91622254e52ad942c6a6d3778b14R71-R73

You are right.  I'm sorry that I missed there is `if initial_bias is None: initial_bias = 0`.  It looks good as it is because `L.Linear` does the same.

toslunar commented on chainer/links/rnn/n_step_rnn.py:
```diff
@@ -51,21 +51,26 @@ class NStepRNNBase(link.ChainList):
+            'Use chainer.using_config',
+            use_bi_direction='use_bi_direction is not supported anymore',
+            activation='activation is not supported anymore',
+        )
```
Optional: The signature could be
```
__init__(self, n_layers, in_size, out_size, dropout, *, initialW=None, initial_bias=None, **kwargs)
```
as Python 2 support has been dropped.

himkt replied:
9fb69b3 0c87915
Updated to use keyword arguments.

toslunar commented on tests/chainer_tests/links_tests/rnn_tests/test_link_n_step_rnn.py:
```diff
@@ -413,4 +414,74 @@ def test_n_cells(self):
+class TestInitialization(unittest.TestCase):
+    def setUp(self):
+        if self.initializer is None:
+            initializer = initializers.constant.Zero()
```
`'zero'` could be more informative than `None`.

himkt replied:
8cc5033
Thanks, I updated.

toslunar commented on tests/chainer_tests/links_tests/rnn_tests/test_link_n_step_rnn.py:
```diff
@@ -413,4 +414,74 @@ def test_n_cells(self):
+        initializer(self.initial_bias)
+
+        # FIXME (himkt) workaround
+        self.initial_bias = self.initial_bias.reshape(-1)
```
Could you explain what issue is fixed by the workaround?

himkt replied:
I put this tricky workaround because NStepRNN accepts a `n`
dimensional ndarray and Initializer accepts `(n, 1)` ndarray.

If there is a good way to remove it, please let me know. :bow:

---

#### When `self.initial_bias = numpy.zeros(10, dtype=self.dtype)`

- Initializer:
`ValueError: shape must be of length >= 2. Actual shape: (10,)`

```python
chainer/initializers/uniform.py:115: in __call__
    fan_in, fan_out = initializer.get_fans(array.shape)
```

#### When `self.initial_bias = numpy.zeros((10, 1), dtype=self.dtype)`

- NStepRNN:
`ValueError: could not broadcast input array from shape (10,1) into shape (10)`

```python
chainer/links/rnn/n_step_rnn.py:85: in __init__
    b = variable.Parameter(bias_initializer, out_size)
chainer/variable.py:1805: in __init__
    data = initializers.generate_array(initializer, shape, xp)  # type: ignore # NOQA
chainer/initializers/__init__.py:74: in generate_array
    initializer(array)
chainer/initializers/constant.py:58: in __call__
    array, self.fill_value.astype(array.dtype, copy=False))
chainer/backend.py:63: in copyto
    numpy.copyto(dst, _cpu._to_cpu(src))
```

toslunar replied:
It seems reasonable to me that `initial_bias=GlorotUniform()` is not allowed.

himkt replied:
e125f80 1544739
Replace GlorotUniform with Uniform in test and prohibit to use them for `initial_bias`.

himkt replied:
e125f80
Update to use `Uniform` instead of `GlorotUniform` for initializing a bias vector.

toslunar commented on tests/chainer_tests/links_tests/rnn_tests/test_link_n_step_rnn.py:
```diff
@@ -413,4 +414,74 @@ def test_n_cells(self):
+    @attr.gpu
+    def test_param_gpu(self):
+        with testing.assert_warns(DeprecationWarning):
+            self.link.to_gpu()
```
`self.link.to_device('@cupy:0')` is preferred to `to_gpu` in new tests.

himkt replied:
Replaced in 0674e07

toslunar commented on tests/chainer_tests/links_tests/rnn_tests/test_link_n_step_rnn.py:
```diff
@@ -413,4 +414,74 @@ def test_n_cells(self):
+        self.initial_bias = numpy.zeros((10, 1), dtype=self.dtype)
+
+        initializer(self.initialW)
+        initializer(self.initial_bias)
```
How about `self.initialW = numpy.random.uniform(-1, 1, (10, 10)).astype(self.dtype)`?

To test initializers, `setUp` should have `self.link = link(..., initialW=initializer, initial_bias=initializer)`.

himkt replied:
1. Updated to initialize `self.initialW` and `self.initial_bias` by `numpy.random.uniform`.

2. The current implementation is based on [this approach](https://github.com/chainer/chainer/pull/5807#discussion_r242031702) since I couldn't find a way to reproduce the same initialization results in `setUp` and `NStepRNN.__init__` (setting seed before/after `initializer(self.initialW)` didn't work). Could you please let me know if you have any ideas?

himkt commented on chainer/links/rnn/n_step_rnn.py:
```diff
@@ -50,22 +50,26 @@ class NStepRNNBase(link.ChainList):
-                activation='activation is not supported anymore')
-            argument.assert_kwargs_empty(kwargs)
+    def __init__(self, n_layers, in_size, out_size, dropout,
+                 initialW=None, initial_bias=None, **kwargs):
```
:bow:

himkt commented on chainer/links/rnn/n_step_rnn.py:
```diff
@@ -50,22 +50,26 @@ class NStepRNNBase(link.ChainList):
 
-    def __init__(self, n_layers, in_size, out_size, dropout, **kwargs):
-        if kwargs:
-            argument.check_unexpected_kwargs(
```
Fixed in d1d3222

himkt replied:
ee5e00f

1. I restored `if kwargs:` since it needs in the point of performance.
2. Although I also restored `assert_kwargs_empty`, I still don't understand why `argument.assert_kwargs_empty` is required (I guess `if kwargs:` check if `kwargs` is empty). I'm sorry but could you please give me some explanations for this?

himkt commented on chainer/links/rnn/n_step_rnn.py:
```diff
@@ -50,22 +50,32 @@ class NStepRNNBase(link.ChainList):
+        if isinstance(bias_initializer,
+                      (initializers.GlorotUniform, initializers.GlorotNormal)):
+            raise ValueError('GlorotUniform and GlorotNormal'
+                             ' are not allowed to initialize bias vector')
```
Oh, yes indeed. I removed in 7548cf1

himkt commented:
https://github.com/chainer/chainer/pull/8489#pullrequestreview-330364093

1.  Updated the first line
2. Updated `Args` to use the same description as them in `Linear`

himkt commented:
babd9f9

Although setting `rng` seems to work well (as below),
passing it to an initializer still produces different initial weights from expected ones.

```python3

In [1]: import copy, chainer, numpy
In [2]: a = chainer.initializers.GlorotUniform(rng=numpy.random.RandomState(seed=1))
In [3]: a2 = copy.deepcopy(a)
In [4]: b = numpy.random.random((10, 10))
In [5]: b2 = numpy.random.random((10, 10))
In [6]: a(b)
In [7]: a2(b2)
In [8]: b - b2
Out[8]:
array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])
```

<details>

```python
In [9]: b
Out[9]:
array([[-0.09089784,  0.24135339, -0.54759727, -0.21653382, -0.38695953,
        -0.44657069, -0.34368472, -0.16917975, -0.11308557,  0.0425216 ],
       [-0.08851797,  0.2028978 , -0.32375634,  0.4142069 , -0.51772095,
         0.1867378 , -0.09058805,  0.06429149, -0.39393637, -0.33071325],
       [ 0.32944917,  0.51295486, -0.20438357,  0.21067887,  0.41231366,
         0.43226994, -0.45456129, -0.50494019, -0.36168265,  0.41423436],
       [-0.439989  , -0.08642227,  0.50159285,  0.03633075,  0.21019085,
        -0.2020925 ,  0.20430153,  0.36656406, -0.52768875,  0.27401937],
       [ 0.53552049,  0.27185185, -0.24051156,  0.31688963, -0.43464413,
        -0.05707978,  0.44759395, -0.22608437, -0.23248047, -0.40528339],
       [-0.52650712,  0.19590451, -0.31589557, -0.25683077, -0.00923114,
        -0.48926682,  0.08119177, -0.38698946,  0.09782931,  0.21882432],
       [-0.43562081, -0.09414695,  0.2129547 , -0.0940119 , -0.49300128,
         0.03932254,  0.17942804,  0.01631021,  0.48702915,  0.0948163 ],
       [ 0.44190466, -0.39712656, -0.39515296,  0.33673029, -0.11208941,
        -0.36658611,  0.46831219, -0.16676415,  0.27475089,  0.24756839],
       [ 0.41989079,  0.13547612,  0.27489366, -0.16552357, -0.25203137,
         0.43367162, -0.07877215,  0.50920676,  0.17904119,  0.13331098],
       [-0.42202464,  0.49239081, -0.05486851,  0.08587152, -0.10063109,
        -0.28807251,  0.44188013,  0.08071183, -0.54457827,  0.12832582]])

In [10]: b2
Out[10]:
array([[-0.09089784,  0.24135339, -0.54759727, -0.21653382, -0.38695953,
        -0.44657069, -0.34368472, -0.16917975, -0.11308557,  0.0425216 ],
       [-0.08851797,  0.2028978 , -0.32375634,  0.4142069 , -0.51772095,
         0.1867378 , -0.09058805,  0.06429149, -0.39393637, -0.33071325],
       [ 0.32944917,  0.51295486, -0.20438357,  0.21067887,  0.41231366,
         0.43226994, -0.45456129, -0.50494019, -0.36168265,  0.41423436],
       [-0.439989  , -0.08642227,  0.50159285,  0.03633075,  0.21019085,
        -0.2020925 ,  0.20430153,  0.36656406, -0.52768875,  0.27401937],
       [ 0.53552049,  0.27185185, -0.24051156,  0.31688963, -0.43464413,
        -0.05707978,  0.44759395, -0.22608437, -0.23248047, -0.40528339],
       [-0.52650712,  0.19590451, -0.31589557, -0.25683077, -0.00923114,
        -0.48926682,  0.08119177, -0.38698946,  0.09782931,  0.21882432],
       [-0.43562081, -0.09414695,  0.2129547 , -0.0940119 , -0.49300128,
         0.03932254,  0.17942804,  0.01631021,  0.48702915,  0.0948163 ],
       [ 0.44190466, -0.39712656, -0.39515296,  0.33673029, -0.11208941,
        -0.36658611,  0.46831219, -0.16676415,  0.27475089,  0.24756839],
       [ 0.41989079,  0.13547612,  0.27489366, -0.16552357, -0.25203137,
         0.43367162, -0.07877215,  0.50920676,  0.17904119,  0.13331098],
       [-0.42202464,  0.49239081, -0.05486851,  0.08587152, -0.10063109,
        -0.28807251,  0.44188013,  0.08071183, -0.54457827,  0.12832582]])
```

</details>

toslunar commented on chainer/links/rnn/n_step_rnn.py:
```diff
@@ -27,7 +27,8 @@ def permutate_list(lst, indices, inv):
 class NStepRNNBase(link.ChainList):
-    """__init__(self, n_layers, in_size, out_size, dropout)
+    """__init__(self, n_layers, in_size, out_size, dropout, \
+*, initialW=None, initial_bias=None, **kwargs):
```
```suggestion
*, initialW=None, initial_bias=None)
```
- Deprecated arguments are usually removed from the signature on the doc.
- `:`

toslunar commented:
> Although setting `rng` seems to work well (as below),
> passing it to an initializer still produces different initial weights from expected ones.

How about this https://github.com/chainer/chainer/compare/himkt:init-rnn...toslunar:init-rnn?

himkt commented:
Thanks, it works.

memo

- NStepRNNBase [uses an initializer repeatedly](https://github.com/chainer/chainer/pull/8489/files#diff-e40e91622254e52ad942c6a6d3778b14R95-R96)
- [Setting seed in the head of `check_param`](https://github.com/chainer/chainer/pull/8489/files#diff-5ae7126a422977a832a43c5094138a7eR466) simulates this behavior.

toslunar commented:
Jenkins & flexCI, test this please.

toslunar commented:
Jenkins, test this please.

toslunar commented:
The Jenkins failures are #8481.
